tensorstore/kvstore/tsgrpc/kvstore_server.cc

```diff
@@ -151,49 +159,101 @@ class ReadHandler final : public Handler<ReadRequest, ReadResponse> {
+    if (!ok) {
+      // OnDone is going to be called after we return from this method.
+      Finish(::grpc::Status(::grpc::StatusCode::UNKNOWN, "Write failed"));
+      return;
+    }
+    if (value_offset_ < value_.size()) {
+      response_.Clear();
+      SetNextPart();
+      StartWrite(&response_);
+    } else {
+      Finish(::grpc::Status::OK);
+    }
+  }
+
+  void SetNextPart() {
+    auto next_part = value_.Subcord(value_offset_, max_sent_part_bytes_);
+    response_.set_value_part(std::move(next_part));
+    value_offset_ = std::min(value_.size(), value_offset_ + next_part.size());
   }
 
  private:
   KvStore kvstore_;
+  size_t max_sent_part_bytes_;
   Future<void> future_;
+
+  ReadResponse response_;
+  absl::Cord value_;
+  size_t value_offset_ = 0;
 };
 
-class WriteHandler final : public Handler<WriteRequest, WriteResponse> {
-  using Base = Handler<WriteRequest, WriteResponse>;
+class WriteHandler final : public ClientStreamHandler<WriteRequest, WriteResponse> {
+  using Base = ClientStreamHandler<WriteRequest, WriteResponse>;
 
  public:
-  WriteHandler(CallbackServerContext* grpc_context, const Request* request,
-               Response* response, KvStore kvstore)
-      : Base(grpc_context, request, response), kvstore_(std::move(kvstore)) {}
+  WriteHandler(CallbackServerContext* grpc_context,
+                      Response* response, KvStore kvstore)
+      : Base(grpc_context, response), kvstore_(std::move(kvstore)) {}
 
   void Run() {
-    ABSL_LOG_IF(INFO, verbose_logging)
-        << "WriteHandler " << ConciseDebugString(*request());
-    tensorstore::kvstore::WriteOptions options{};
-    options.if_equal.value = request()->generation_if_equal();
+    StartRead(&request_);
+  }
 
-    internal::IntrusivePtr<WriteHandler> self{this};
-    future_ =
-        PromiseFuturePair<void>::Link(
-            [self = std::move(self)](Promise<void> promise, auto write_result) {
-              if (!promise.result_needed()) return;
-              promise.SetResult(self->HandleResult(write_result.result()));
-            },
-            kvstore::Write(kvstore_, request()->key(),
-                           absl::Cord(request()->value()), options))
-            .future;
+  void OnReadDone(bool ok) override {
+    if (!ok) {
+      if (grpc_context()->IsCancelled()) {
+        // OnCancelled is going to be called and will invoke Finish, thus just stop reading sequence here.
+        return;
+      } else {
+        // Perform actual writing.
+        internal::IntrusivePtr<WriteHandler> self{this};
+        future_ =
+          PromiseFuturePair<void>::Link(
+              [self = std::move(self)](Promise<void> promise, auto write_result) {
+                if (!promise.result_needed()) return;
+                // Actual value set to the promise does not matter, as the promise is only used for tracking completion
+                // or cancellation.
+                promise.SetResult(absl::OkStatus());
```

Comment:
I think that this should be more like:

promise.SetResult(self->HandleResult(write_result.result()));

Or at least, the HandleResult() needs the status response handled correctly.